FIX EagerLoading: Don't throw unknown column error when list is empty

beerbohmdo opened the pull request:
<!--
  Thanks for contributing, you're awesome! ⭐

  Please read https://docs.silverstripe.org/en/contributing/code/ if you haven't contributed to this project recently.
-->
## Description
<!--
  Please describe expected and observed behaviour, and what you're fixing.
  For visual fixes, please include tested browsers and screenshots.
-->
The EagerLoadedList currently throws an error when the list is empty and you try to filter it.

## Manual testing steps
<!--
  Include any manual testing steps here which a reviewer can perform to validate your pull request works correctly.
  Note that this DOES NOT replace unit or end-to-end tests.
-->

## Issues
<!--
  List all issues here that this pull request fixes/resolves.
  If there is no issue already, create a new one! You must link your pull request to at least one issue.
-->
- #11893

## Pull request checklist
<!--
  PLEASE check each of these to ensure you have done everything you need to do!
  If there's something in this list you need help with, please ask so that we can assist you.
-->
- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [x] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

beerbohmdo committed:
FIX Don't throw unknown column error when list is empty

GuySartorelli reviewed (changes requested):
This PR feels like it's working around a problem that actually lies in `getMatches()` where the exception is being thrown from. Is there a way to resolve this in that method instead of in each method that calls it?

beerbohmdo commented:
The error is thrown in `normaliseFilterArgs` not in `getMatches`. And I can't really change that. I could only change the return value of that method.  I could just return an empty filter or all fields, but both just feels wrong. I could return an distinct value like null in that case, but then I would have to change the calling methods also.

beerbohmdo commented:
@GuySartorelli  I would like to upgrade my PR to ss 6, but for that we should answer the question howto resolve the issue.

lekoala commented:
hi @GuySartorelli and @beerbohmdo . i just got hit recently by this and it's very surprising because it doesn't trigger on regular workload

so this: 
```php
$records = $myDataList->eagerLoad('MyRelation.NestedRelation')
```
may work or may fail (at least, on my setup in 5.4 but it seems exactly the same cause)

i think this PR has the benefit to fix the issue and cover it with a test case. it doesn't mean it cannot be improved by a follow up ticket. at any rate, the current state basically prevents to use this in a reliable fashion.

or, alternate approach, add a new private method like

```php
private function getMatchesFromArgs(array $args, string $function, bool $any = false): array
{
    if (empty($this->rows)) {
        return [];
    }

    $filters = $this->normaliseFilterArgs($args, $function);

    return $this->getMatches($filters, $any);
}
```

and

```php
public function filter(...$args): static
{
    $list = clone $this;
    $list->rows = $this->getMatchesFromArgs($args, __FUNCTION__);
    return $list;
}
```

(not tested, just a proposal to make this move forward since I had to check it anyways...)

beerbohmdo commented:
Found an additional location where eager loading may breaks because of an empty list.

If you have a multi level relation like this:

`MyDataObject::get()->eagerLoad('Products', 'Products.Categories')`

And the `Products` relation is empty, the `Categories` will throw an error like this:

```
[Emergency] Uncaught InvalidArgumentException: Cannot filter "Product_Categories"."ProductID" against an empty set
```

The issue is that the root collection is checked here:

https://github.com/silverstripe/silverstripe-framework/blob/6/src/ORM/DataList.php#L1150

but the sub loop https://github.com/silverstripe/silverstripe-framework/blob/6/src/ORM/DataList.php#L1171 does not check for an empty $parentIDs array again...

beerbohmdo committed:
FIX Gracefully handle empty relation lists when eager loading

Eager loading a multi-level relation chain where the first level relation is empty for every record no longer tries to fetch the sub-relation against an empty set of parent IDs.

Filtering, filterAny, exclude or excludeAny on an empty eager-loaded list now returns an empty result instead of throwing "Can't filter by column".

Fixes #11893

lekoala commented:
my hero :)
